pkg/operator/controller/machineconfig/controller.go

```diff
@@ -0,0 +1,354 @@
+	withoutFinalizers := r.CtrlConfig.DeepCopy()
+
+	newFinalizers := make([]string, 0)
+	for _, item := range withoutFinalizers.Finalizers {
+		if item == finalizer {
+			continue
+		}
+		newFinalizers = append(newFinalizers, item)
+	}
+	if len(newFinalizers) == 0 {
+		// Sanitize for unit tests, so we don't need to distinguish empty array
+		// and nil.
+		newFinalizers = nil
+	}
+	withoutFinalizers.Finalizers = newFinalizers
+	if err := r.Update(ctx, withoutFinalizers); err != nil {
+		return withoutFinalizers, err
+	}
+	return withoutFinalizers, nil
+}
+
+func hasFinalizer(mc *v1alpha1.Machineconfig) bool {
+	hasFinalizer := false
+	for _, f := range mc.Finalizers {
+		if f == finalizer {
+			hasFinalizer = true
+			break
+		}
+	}
+	return hasFinalizer
+}
+
+// checkProfConf checks and ensures profiling config for defined services
+func (r *MachineconfigReconciler) checkProfConf(ctx context.Context) error {
+	r.Lock()
+	defer r.Unlock()
+
+	errored := false
+	errs := fmt.Errorf("failed to check profiling configs")
+	if r.CtrlConfig.Spec.EnableCrioProfiling {
+		criomc, created, err := r.ensureCrioProfConfigExists(ctx)
+		if err != nil {
+			errored = true
+			r.Log.Error(err, "failed to enable crio profiling")
+			errs = fmt.Errorf("%w: %s", errs, err)
+		}
+		if created {
+			r.EventRecorder.Eventf(r.CtrlConfig, corev1.EventTypeNormal, "CreateConfig", "successfully created crio machine config")
+			r.PrevSyncChange["crio"] = PrevSyncData{
+				action: "created",
+				config: *criomc,
+			}
+		}
+	} else {
+		deleted, err := r.ensureCrioProfConfigNotExists(ctx)
+		if err != nil {
+			errored = true
+			r.Log.Error(err, "failed to disable crio profiling")
+			errs = fmt.Errorf("%w: %s", errs, err)
+		}
+		if deleted {
+			r.EventRecorder.Eventf(r.CtrlConfig, corev1.EventTypeNormal, "DeleteConfig", "successfully deleted crio machine config")
+			r.PrevSyncChange["crio"] = PrevSyncData{
+				action: "deleted",
+			}
+		}
+	}
+
+	if r.CtrlConfig.Spec.EnableKubeletProfiling {
+		kubeletmc, created, err := r.ensureKubeletProfConfigExists(ctx)
+		if err != nil {
+			errored = true
+			r.Log.Error(err, "failed to enable kubelet profiling")
+			errs = fmt.Errorf("%w: %s", errs, err)
+		}
+		if created {
+			r.EventRecorder.Eventf(r.CtrlConfig, corev1.EventTypeNormal, "CreateConfig", "successfully created kubelet config")
+			r.PrevSyncChange["kubelet"] = PrevSyncData{
+				action: "created",
+				config: *kubeletmc,
+			}
+		}
+	} else {
+		deleted, err := r.ensureKubeletProfConfigNotExists(ctx)
+		if err != nil {
+			errored = true
+			r.Log.Error(err, "failed to disable kubelet profiling")
+			errs = fmt.Errorf("%w: %s", errs, err)
+		}
+		if deleted {
+			r.EventRecorder.Eventf(r.CtrlConfig, corev1.EventTypeNormal, "DeleteConfig", "successfully deleted kubelet config")
+			r.PrevSyncChange["kubelet"] = PrevSyncData{
+				action: "deleted",
+			}
+		}
+	}
+
+	if errored {
+		return errs
+	}
+	return nil
```

Comment:
you can simplify this as
```
import utilerrors "k8s.io/apimachinery/pkg/util/errors"
var errors []error
if err != nil {
    errors = append(errors, err)
}
return utilerrors.NewAggregate(errors)
```

Reply:
comment incorporated in [d0805fd](https://github.com/openshift/node-observability-operator/pull/16/commits/d0805fd9c9a88344d99aa95ea52cf86ac8ad105c)